File: frontend/web_app/lib/models/category.dart
class Category {
  final String id;
  final String budgetId;
  final String categoryGroupId;
  final String categoryNameId;
  final String categoryName;
  final String activity;
  final String balance;
  final String budgeted;
  final String monthId;
  final bool deleted;
  final bool hidden;

  Category({
    required this.id,
    required this.budgetId,
    required this.categoryGroupId,
    required this.categoryNameId,
    required this.categoryName,
    required this.activity,
    required this.balance,
    required this.budgeted,
    required this.monthId,
    required this.deleted,
    required this.hidden,
  });

  factory Category.fromJson(Map<String, dynamic> json) {
    return Category(
      id: json['id'],
      budgetId: json['budget_id'],
      categoryGroupId: json['category_group_id'],
      categoryNameId: json['category_name_id'],
      categoryName: json['category_name'],
      activity: json['activity'],
      balance: json['balance'],
      budgeted: json['budgeted'],
      monthId: json['month_id'],
      deleted: json['deleted'] ?? false,
      hidden: json['hidden'] ?? false,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'budget_id': budgetId,
        'category_group_id': categoryGroupId,
        'category_name_id': categoryNameId,
        'category_name': categoryName,
        'activity': activity,
        'balance': balance,
        'budgeted': budgeted,
        'month_id': monthId,
        'deleted': deleted,
        'hidden': hidden,
      };
}

File: backend/Python/models/category.py
#!/usr/bin/env python3
from models.base import BaseModel, db, uuid4, UUID

# -------------------------------
# Category Model
# -------------------------------

class Category(BaseModel):
    """
    Represents a transaction category for a specific month.
    """
    __tablename__ = 'categories'

    id = db.Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    category_name_id = db.Column(db.String(36), db.ForeignKey('category_names.id'), nullable=False)
    category_group_id = db.Column(db.String(36), db.ForeignKey('category_groups.id'), nullable=False)
    hidden = db.Column(db.Boolean, default=False)
    deleted = db.Column(db.Boolean, default=False)
    budget_id = db.Column(db.String(36), db.ForeignKey('budgets.id'), nullable=False)
    budgeted = db.Column(db.Numeric(15, 2), default=0)
    activity = db.Column(db.Numeric(15, 2), default=0)
    balance = db.Column(db.Numeric(15, 2), default=0)
    month_id = db.Column(db.String(36), db.ForeignKey('months.id'), nullable=False)

    transactions = db.relationship('Transaction', back_populates='category')
    category_group = db.relationship('CategoryGroup', back_populates='categories')
    month = db.relationship('Month', back_populates='categories')
    category_name = db.relationship('CategoryName', back_populates='categories')

class CategoryName(BaseModel):
    """
    Represents the category name (normalized across months).
    """
    __tablename__ = 'category_names'

    id = db.Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    name = db.Column(db.Text, nullable=False, unique=True)

    categories = db.relationship('Category', back_populates='category_name')

# -------------------------------
# CategoryGroup Model
# -------------------------------

class CategoryGroup(BaseModel):
    """
    Represents a group of categories.
    """
    __tablename__ = 'category_groups'

    id = db.Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    name = db.Column(db.Text, nullable=False, unique=True)
    deleted = db.Column(db.Boolean, default=False)
    budget_id = db.Column(db.String(36), db.ForeignKey('budgets.id'), nullable=False)

    categories = db.relationship('Category', back_populates='category_group')